### ISSUE TITLE
TS2322: Property 'css' does not exist...

### ISSUE THREAD

**Issue by peterschussheim** [emotion-js/emotion#1249]
- `emotion`: version: 10.0.7
- `emotion-theming`: 10.0.7
- `@emotion/core`: version: 10.0.7
- `@emotion/styled`:  version: 10.0.7
- `@emotion/cache`:  version: 10.0.
- `react` version: 16.7.0
- `typescript` version: 3.2.4

## **Relevant code:**

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "allowUnreachableCode": false,
    "checkJs": false,
    "downlevelIteration": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "jsx": "preserve",
    "lib": ["webworker", "esnext", "dom"],
    "module": "esnext",
    "moduleResolution": "node",
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": false,
    "noImplicitReturns": true,
    "noImplicitThis": false,
    "noUnusedLocals": false,
    "outDir": "./build/",
    "pretty": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": false,
    "strictNullChecks": false,
    "suppressImplicitAnyIndexErrors": true,
    "target": "esnext",
    "types": ["@emotion/core", "jest", "node", "webpack-env"]
  },
  "include": ["./src/**/*", "./monaco.d.ts"],
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "gamma.config.js"
  ]
}

```

## Example TS error:

```bash
 TS2322: Type '{ children: Element; css: SerializedStyles; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
Property 'css' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
```

### pragma

```typescript
/** @jsx jsx */
import { jsx, css } from '@emotion/core'
import * as React from 'react'
...
```

## Related Issues:

- https://github.com/emotion-js/emotion/issues/693

**Comment by nhooyr** [comment]
I fixed this by changing my tsconfig.json to:

```javascript
{
  "compilerOptions": {
    "types": ["@emotion/core"]
  }
}
```

**Comment by andrewlsimplisafe** [comment]
Ran into the same issue. If you have already tried @nhooyr 's solution and still get errors. The next step would be a fresh install of dependencies.  (npm install or yarn install). If that doesn't work, try nuking (deleting) the node_modules folder and do a fresh reinstall then reload your IDE / text editor

**Comment by kingerez** [comment]
I'm still getting the same error, none of the above helped... I'm have a freshly created an app using
`npx create-react-app --typescript`
I made the change to tsconfig.json, completely deleted node_modules, reinstalled and restarted my IDE, without success.

**Comment by havenchyk** [comment]
If it helps somehow, you can create custom type declaration like next one


```
// @types/react/index.d.ts
declare module "react" {
  interface Attributes {
      css?: any;
  }
}
```

and specify in `typeRoots: ['./@types', './node_modules/@types']`, order matters!

**Comment by luisrodriguez-cbsinteractive** [comment]
I keep getting the not found error even thought the definitions is in the same file.
I converting my js files to tsx typescript as used in my office

/Users/lrodr0923/repos/brackets/src/App.tsx
TypeScript error in /Users/lrodr0923/repos/brackets/src/App.tsx(34,6):
Type '{ teamName: string; coachName: string; cardDate: string; teamDescription: string; actions: { id: number; label: Element; }[]; }' is not assignable to type 'IntrinsicAttributes & IBrackets'.
  Property 'teamName' does not exist on type 'IntrinsicAttributes & IBrackets'.  TS2322

    32 | 
    33 |   return (
  > 34 |     <Card 
       |      ^
    35 |       teamName="{teamRecord.teamName}"
                             KEEPS TELLING ME, THE VARIABLE IS NOT FOUND, but the definition is there.
    36 |       coachName={teamRecord.coachName}
    37 |       cardDate={cardDate}

and this is what the interface definition looks like:
interface IBracket {
    id: number;
    logoUrl: string,
    teamName: string,
    coach: string,
    season: string,
    seasonDetails: {
      overallRecord: string,
      gamesWon: number,
      gamesLost: number,
    }
    playersLost:[
      {
        firstname: string
        lastname: string
        age: number
        position: ""
      }
    ]
    playersGained:[
      {
        firstname: string
        lastname: string
        age: number
        position: ""
      }
    ]
}

interface IBrackets {
    brackets: IBracket[];
}

Thanks for helping

**Comment by MarcStd** [comment]
Are there any updates on this issue? I am still getting this problem. Is a special tsconfig required?

**Comment by MegaCookie** [comment]
For emotion 11 use the following configuration in your `tsconfig.json`:
```json
{
  "compilerOptions": {
    "types": ["@emotion/react/types/css-prop"]
  }
}
```

**Comment by floroz** [comment]
> For Emotion 11 use the following configuration in your `tsconfig.json` until TypeScript 4.1 comes out:
> 
> ```json
> {
>   "compilerOptions": {
>     "types": ["@emotion/react/types/css-prop"]
>   }
> }
> ```
> 
> Documentation: https://emotion.sh/docs/emotion-11#css-prop-types

thanks that worked for me!

**Comment by Vadorequest** [comment]
For me, it was adding `/// <reference types="@emotion/react/types/css-prop" />` to `next-env.d.ts` file. See https://emotion.sh/docs/emotion-11#css-prop-types

See https://github.com/UnlyEd/next-right-now/commit/2a3c89dad00cf212929a54b8d4bb2d14f0677af5

**Comment by saulpalv** [comment]
just add the following line to react-app-env.d.ts as stated in https://emotion.sh/docs/emotion-11
`/// <reference types="@emotion/react/types/css-prop" /> `

and import with

```
/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react'
```

**Comment by santyclaz** [comment]
**TL;DR:** for **React 17, Typescript 4.2, Emotion 11.1**

`tsconfig.json`
```js
{
  "compilerOptions": {
    ...
    "jsx": "react-jsx",
    "jsxImportSource": "@emotion/react",
    ...
  }
}
```

This seems to be the cleanest solution that works for me that does NOT require:
1. adding any `/// <reference types="@emotion/react/types/css-prop" />` to a `.d.ts` file
2. nor adding a `/** @jsxImportSource @emotion/react */` with any imports

See: [Emotion TypeScript `css` prop docs](https://emotion.sh/docs/typescript#css-prop)

-------------

Additional Context:

Prior comments and the [Emotion 11 docs](https://emotion.sh/docs/emotion-11#css-prop-types) linked in prior comments  seemed to suggest that upgrading to a version of TypeScript > 4.1 would somehow resolve this issue (and by "resolve" I presumed this would mean removing the need for either the suggested `tsconfig.json` changes, or the other change suggested by a few others above on adding a `/// <reference types="@emotion/react/types/css-prop" />` line in some `.d.ts` file in your codebase. Unfortunately this doesn't seem to be case.

**Comment by holylander** [comment]
Hi guys,
in my case it was a matter of first, updating typesdcript config so it allows such prop:
```
{
  "compilerOptions": {
    "types": ["@emotion/react/types/css-prop"]
  }
}
```

and then update my babel config so it transpiles the styles declared in the prop:
```
presets: [
                            [
                                '@babel/preset-react',
                                {
                                    // https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#manual-babel-setup
                                    runtime: 'automatic',
                                    importSource: "@emotion/react"
                                }
                            ]
                        ],

                        plugins: [
                            '@emotion',
                        ],
]
```

**Comment by RV12R** [comment]
Add `{
  "compilerOptions": {
    "types": ["@emotion/react/types/css-prop"]
  }
}`  to `tsconfig.json`
And add `/// <reference types="@emotion/react/types/css-prop" />` to `next-env.d.ts`

**Comment by RV12R** [comment]
Or use this `npm install --save @emotion/css`

**Comment by kaguya3222** [comment]
Why is this closed? I have issues with this

**Comment by SRochas** [comment]
With TypeScript@4.9.4, React@18.2.0, @emotion/react@11.10.5, TypeScript in strict mode:

My use-case is slightly different as I'm building components whose types extend the HTMLAttributes of basic types. Before upgrading @emotion/react (from @emotion/react@11.1.5), the `css` prop was included in my type when I extended the HTMLAttributes for the given type. After the upgrade, I had to convert the following:

```
export interface TextboxProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
...

const Textbox = ({ 
    css, // TS2322: Property `css` does not exist
    ...
```

to include the `css` prop manually. I guess the prop is no longer found in `React.InputHTMLAttributes<HTMLInputElement>` anymore, or perhaps something was misconfigured in my solution before the upgrade and the upgrade exposed the issue? 

Regardless, this code fixed my issue:

```
import type { CSSObject } from '@emotion/react';

export interface TextboxProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'css'> {
    css?: CSSObject;
...

const Textbox = ({ 
    css, // Fixed
    ...
```

**Comment by Andarist** [comment]
@SRochas Please always try to share a repro case in a runnable form - either by providing a git repository to clone or a codesandbox. OSS maintainers usually can't afford the time to set up the repro, even if exact steps are given.

**Comment by DigitalAgency711** [comment]
```
{
  "compilerOptions": {
    "types": ["@emotion/react/types/css-prop"]
  }
}
```
This is work for me without error.

**Comment by Risyandi** [comment]
> For Emotion 11 use the following configuration in your `tsconfig.json` until TypeScript 4.1 comes out:
> 
> ```json
> {
>   "compilerOptions": {
>     "types": ["@emotion/react/types/css-prop"]
>   }
> }
> ```
> 
> Documentation: https://emotion.sh/docs/emotion-11#css-prop-types

good solutions by adding this config currently working fine.

**Comment by takayui0921** [comment]
just add.    /** @jsxImportSource @emotion/react */ 　to  the component that's having this error.  
it worked for me! no need to change your tsconfig.json.

**Comment by NicholasTaylor** [comment]
I know this is closed, but just leaving a helpful note for anyone that might come across this. If your linter is complaining about this error, but your site loads fine while running `npm run dev`, your main `tsconfig.json` might be configured right, but the linter's config might need similar tweaks. 

First, check if that's the case by opening `tsconfig.json`. Look for a property called `references`. If it looks something like this, you may have a little more work left to do:
```
"references": [
  { "path": "./tsconfig.app.json" },
  { "path": "./tsconfig.node.json" }
]
```
Open `tsconfig.app.json`. There should be a section of this json with a header comment saying `/* Linting*/` Anywhere in this section, just put in the following property:
`"jsxImportSource": "@emotion/react",`

Then, just save and reload your IDE. The linter should stop complaining at that point.
